File: Pc191/LeetCode/Tree/construct_tree.py
# from collections import deque
# class Node:
#     def __init__(self,data):
#         self.data=data
#         self.left=None
#         self.right=None

# class Tree:
#     def __init__(self):
#         self.root=None
    
#     def isEmpty(self):
#         if not self.root:
#             return True
#         return False
    
#     def countNode(self,root):
#         if not root:
#             return 0
        
#         return 1+self.countNode(root.left)+self.countNode(root.right)
    
#     def height(self,root):
#         if root is None:
#             return 0
        
#         return max(self.height(root.left),self.height(root.right))+1
    
#     def preorder(self,root):
#         if root is None:
#             return
#         print(root.data,end=' ')
#         self.preorder(root.left)
#         self.preorder(root.right)
    
#     def inorder(self,root):
#         if root is None:
#             return
#         self.inorder(root.left)
#         print(root.data,end=' ')
#         self.inorder(root.right)
        
    
#     def postorder(self,root):
#         if root is None:
#             return
#         self.postorder(root.left)
#         self.postorder(root.right)
#         print(root.data,end=' ')
    
#     def bfs(self,root):
#         if root is None:
#             return
#         q=deque()
#         q.append(root)
#         while q:
#             node=q.popleft()
#             print(node.data,end=" ")
            
#             if node.left:
#                 q.append(node.left)
#             if node.right:
#                 q.append(node.right)
            
        
        
        
        
    
        
# if __name__=='__main__':
#     tree=Tree()
#     tree.root=Node(1)
#     tree.root.left=Node(2)
#     tree.root.right=Node(3)
#     tree.root.left.left=Node(4)
#     tree.root.left.right=Node(5)
#     tree.root.right.left=Node(6)
#     tree.root.right.right=Node(7)
#     print(tree.isEmpty())
#     print(tree.countNode(tree.root))
    
#     print(tree.height(tree.root))
    
#     tree.preorder(tree.root)
#     print()
#     tree.inorder(tree.root)
#     print()
#     tree.bfs(tree.root)
    

from collections import deque
    
class Node:
    def __init__(self,data):
        self.data=data
        self.left=None
        self.right=None

class Tree:
    def __init__(self):
        self.root=None
    
    def countNodes(self,root):
        if root is None:
            return 0
        return 1+self.countNodes(root.left)+self.countNodes(root.right)
    
    def height(self,root):
        if root is None:
            return 0
        return 1+max(self.height(root.left),self.height(root.right))
    
    def preOrder(self,root):
        if root is None:
            return
        print(root.data,end=" ")
        self.preOrder(root.left)
        self.preOrder(root.right)
    
    def postOrder(self,root):
        if root is None:
            return
        
        self.postOrder(root.left)
        self.postOrder(root.right)
        print(root.data,end=" ")

    def inOrder(self,root):
        if root is None:
            return 
        self.inOrder(root.left)
        print(root.data,end=" ")
        self.inOrder(root.right)
    
    
    def bfs(self,root):
        if root is None:
            return
        q=deque()
        q.append(root)
        
        while q:
            node=q.popleft()
            print(node.data,end=" ")
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
    
                
        



if __name__=='__main__':
    tree=Tree()
    node=Node(1)
    tree.root=node
    tree.root.left=Node(2)
    tree.root.right=Node(3)
    tree.root.left.left=Node(4)
    tree.root.left.right=Node(5)
    tree.root.right.left=Node(6)
    tree.root.right.right=Node(7)
    
    print(tree.countNodes(tree.root))
    
    
    

File: Pc191/LeetCode/Tree/construct_tree.cpp
// #include<bits/stdc++.h>
// using namespace std;

// class Node{
    
//     public:
//         int data;
//         Node *left;
//         Node *right;

//         Node(int val){
//             this->data=val;
//             left=NULL;
//             right=NULL;
//         }
    
//         bool isEmpty(Node*root){
//             return root==NULL;
//         }

//         int countNode(Node * root){
//             if(root==NULL){
//                 return 0;

//             }
//             return 1+countNode(root->left)+countNode(root->right);
//         }

//         int height(Node * root){
//             if(root==NULL){
//                 return 0;
//             }

//             return max(height(root->left),height(root->right))+1;
//         }

//         void preorder(Node * root){
//             if(root==NULL){
//                 return;

//             }

//             cout<<root->data<<" ";
//             preorder(root->left);
//             preorder(root->right);

//         }


// };

// int main(){

//     Node* root=new Node(1);
//     root->left=new Node(2);
//     root->right=new Node(3);
//     root->left->left=new Node(4);
//     root->left->right=new Node(5);
//     root->right->left=new Node(6);
//     root->right->right=new Node(7);
//     cout<<root->countNode(root);
//     root->preorder(root);





//     return 0;
// }

#include<bits/stdc++.h>
#include<queue>
using namespace std;

class Node{
    public:
        int data;
        Node* left;
        Node* right;
        
        Node(int data){
            this->data=data;
            this->left=NULL;
            this->right=NULL;

        }

};


class Tree{
    public:
        Node* root;
        Tree(){
            this->root=NULL;

        }

        int countNodes(Node*root){
            if (root == NULL){
                return 0;
            }
            return 1+countNodes(root->left)+countNodes(root->right);
        }

        int height(Node* root){
            if(root == NULL){
                return 0;
            }

            return 1+max(height(root->left),height(root->right));

        }

        void preOrder(Node * root){
            if(root==NULL){
                return;
            }
            cout<<root->data<<" ";
            preOrder(root->left);
            preOrder(root->right);
        }

        void inOrder(Node * root){
            if(root==NULL){
                return;
            }
            
            inOrder(root->left);
            cout<<root->data<<" ";
            inOrder(root->right);
        }

        void postOrder(Node * root){
            if(root==NULL){
                return;
            }
            
            postOrder(root->left);
            postOrder(root->right);
            cout<<root->data<<" ";
        }

        void bfs(Node* root){
           queue<Node *>q;
           q.push(root);
           while(!q.empty()){
            Node * node=q.front();
            q.pop();
            cout<<node->data<<" ";

            if(node->left)q.push(node->left);
            if(node->right)q.push(node->right);

           }

        }
    

};

int main(){

    Tree tree=Tree();
    Node * newNode=new Node(1);
    tree.root=newNode;



    return 0;
}